framework/src/org/checkerframework/common/value/Range.java

```diff
@@ -0,0 +1,670 @@
+        }
+    }
+
+    /**
+     * Returns a range that includes all possible values resulting from signed right shifting an
+     * arbitrary value in this range by an arbitrary number of bits in the specified range. We call
+     * this the signed right shift operation of a range.
+     *
+     * @param right the range of bits by which this range is signed right shifted
+     * @return the range resulting from signed right shifting this range by the specified range
+     */
+    public Range signedShiftRight(Range right) {
+        if (this.isNothing() || right.isNothing()) {
+            return NOTHING;
+        }
+
+        if (this.isWithinInteger() && right.from >= 0 && right.to <= 31) {
+            // The long type can handle the cases with shift distance no longer than 31
+            long resultFrom = from >> (from >= 0 ? right.to : right.from);
+            long resultTo = to >> (to >= 0 ? right.from : right.to);
+            return new Range(resultFrom, resultTo);
+        } else {
+            // Signed shift right operation for long type cannot be simulated with BigInteger.
+            // Give up the calculation and return EVERYTHING instead.
+            return EVERYTHING;
+        }
+    }
+
+    /**
+     * Returns the range of a variable that falls within this range after applying the unary plus
+     * operation (which is a no-op).
+     *
+     * @return this range
+     */
+    public Range unaryPlus() {
+        return this;
+    }
+
+    /**
+     * Returns the range of a variable that falls within this range after applying unary minus
+     * operation.
+     *
+     * @return the resulted range of applying unary minus on an arbitrary value in this range
+     */
+    public Range unaryMinus() {
+        if (this.isNothing()) {
+            return NOTHING;
+        }
+
+        if (from == Long.MIN_VALUE && from != to) {
+            // the only case that needs special handling because of overflow
+            return EVERYTHING;
+        }
+
+        return new Range(-to, -from);
+    }
+
+    /**
+     * Returns the range of a variable that falls within this range after applying bitwise
+     * complement operation.
+     *
+     * @return the resulted range of applying bitwise complement on an arbitrary value in this range
+     */
+    public Range bitwiseComplement() {
+        if (this.isNothing()) {
+            return NOTHING;
+        }
+
+        return new Range(~to, ~from);
+    }
+
+    /**
+     * Refines this range to reflect that some value in it can be less than a value in the given
+     * range. This is used for calculating the control-flow-refined result of the &lt; operator. For
+     * example:
+     *
+     * <pre>
+     * <code>
+     *     {@literal @}IntRange(from = 0, to = 10) int a;
+     *     {@literal @}IntRange(from = 3, to = 7) int b;
+     *     ...
+     *     if (a &lt; b) {
+     *         // range of <i>a</i> is now refined to [0, 6] because a value in range [7, 10]
```

Comment:
In all routines like `refineLessThan`, it seems that the range of `b` should be refined too.  In order to prevent confusion among readers, please discuss how that is handled.

Also, this comment used to be:
```java
     *         // range of <i>a</i> is now refined to [5, 10] because a value in range [0, 4]
     *         // cannot be greater than or equal to variable <i>b</i> with range [3, 5]
```
The (old) comment looks wrong.  In particular, the values 4 could be greater than or equal to values in the other range.  Can you re-check all the comments and the calculations, before I take another look?  Also, note that I changed the exmaples, because using the range [3, 5] makes it confusing, when the result has a 4 in it, whether it is 1 more than the 3 or 1 less than the 5.  The new ranges I used in the documentation are all of extent at least 4 to avoid this problem; more specifically, no 2 values used in the examplse should be within 2 of one another.